File: 初级算法/JavaScript/List/RemoveNthFromEnd.js
/**
 * 删除节点
 * source: https://leetcode-cn.com/explore/interview/card/top-interview-questions-easy/6/linked-list/41/
 */

var deleteNode = function(node) {
	node.val = node.next.val;
	node.next = node.next.next;
};

function ListNode(val) {
	this.val = val;
	this.next = null;
	this.addNode = addNode;
}

function addNode(node) {
	this.next = node;
}

var node0 = new ListNode(0);
var node1 = new ListNode(1);
var node2 = new ListNode(2);
node0.addNode(node1);
node1.addNode(node2);

console.log(node0);
deleteNode(node1)
console.log(node0);

File: 初级算法/Swift/List/removeNthFromEnd.swift
/**
 * 删除链表的倒数第N个节点
 * source: https://leetcode-cn.com/explore/interview/card/top-interview-questions-easy/6/linked-list/42/
 */

import Foundation

class Solution {
    func removeNthFromEnd(_ head: ListNode?, _ n: Int) -> ListNode? {
        
        let dummyNode = ListNode.init(0)
        dummyNode.next = head
        
        var slowNode: ListNode? = dummyNode
        var fastNode = head
        
        for _ in 0..<n {
            fastNode = fastNode?.next
        }
        
        while fastNode != nil {
            fastNode = fastNode?.next
            slowNode = slowNode?.next
        }
        
        slowNode?.next = slowNode?.next?.next
        
        return dummyNode.next
    }
}
